### ISSUE TITLE
[gdb] `02-fused-softmax.py` tutorial hangs when trying to debug

### ISSUE THREAD

**Issue by anmyachev** [intel/intel-xpu-backend-for-triton#5245]
To reproduce:
* `conda create -n triton-xpu python=3.10 && conda activate triton-xpu`
* `./scripts/compile-triton.sh`
* `gh auth login`
* `./scripts/install-pytorch.sh`
* `source /opt/intel/oneapi/setvars.sh`
* Enable kernel debug mode: `find /sys/class/drm/card*/device/ -name "*enable_eu*debug" -exec sh -c 'echo 1 | sudo tee "$1"' _ {} \;`. More details in https://www.intel.com/content/www/us/en/docs/distribution-for-gdb/get-started-guide-linux/2025-2/overview.html
* Run tutorial: `TRITON_INTEL_DISABLE_IGC_OPT=1 DISABLE_LLVM_OPT=1 ZET_ENABLE_PROGRAM_DEBUGGING=1 TRITON_DEBUG=1 LLVM_EXTRACT_DI_LOCAL_VARIABLES=1 /opt/intel/oneapi/2025.2/bin/gdb-oneapi --args python python/tutorials/02-fused-softmax.py`. If put prints in the tutorial sources, one will see that the creation of the tensor by Torch hangs (even if reduce the size to `(32, 16)`): https://github.com/intel/intel-xpu-backend-for-triton/blob/e80a21c818b25b9d6462a0096681092e4dc4790f/python/tutorials/02-fused-softmax.py#L191

**Comment by vlad-penkin** [comment]
Env details:

```
LIBIGC1_VERSION=2.11.12-1146
LEVEL_ZERO_VERSION=1.21.9.0-1136
AGAMA_VERSION=1146
GPU_DEVICE=Intel(R) Data Center GPU Max 1100
```

**Comment by etiotto** [comment]
@anmyachev from your description the test fails because PyTorch hangs in the creation of `x = torch.randn(1823, 781, device=DEVICE) `. 

Given this symptom this is not a Triton issue because the Triton compiler is "responsible" only for the code in the kernel function and the creation of that tensor happens on the host side (outside the kernel). 

We should close this issue. given that the problem is not in the Triton compiler. 

Also what is blocking this issue ?

**Comment by anmyachev** [comment]
> Given this symptom this is not a Triton issue because the Triton compiler is "responsible" only for the code in the kernel function and the creation of that tensor happens on the host side (outside the kernel).

I agree that this is not a bug on Triton's side, but we would still like to track gdb support for such cases. 

> Also what is blocking this issue ?

As far as I understand, we are blocked by a bug fix on the pytorch/gdb side. @vlad-penkin any updates by chance?

**Comment by ghost** [comment]
@anmyachev, this should be fixed by Intel Distribution for GDB (gdb-oneapi) release 2025.3.1. Please try again when the release is available.

**Comment by radocaj-int** [comment]
@anmyachev  the problem that u reported is fixed in latest 2026.0 drop: 

(triton_debug) gta@gkldtt-dev-024:~/intel-xpu-backend-for-triton$ TRITON_INTEL_DISABLE_IGC_OPT=1 DISABLE_LLVM_OPT=1 ZET_ENABLE_PROGRAM_DEBUGGING=1 TRITON_DEBUG=1 LLVM_EXTRACT_DI_LOCAL_VARIABLES=1 /opt/intel/oneapi/2026.0/bin/gdb-oneapi --args python python/tutorials/02-fused-softmax.py
GNU gdb (Intel(R) Distribution for GDB* 2026.0.0) 17.1
Copyright (C) 2026 Free Software Foundation, Inc.; (C) 2026 Intel Corp.
License GPLv3+: GNU GPL version 3 or later <[http://gnu.org/licenses/gpl.html>](http://gnu.org/licenses/gpl.html%3E)
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Type "show copying" and "show warranty" for details.
This GDB was configured as "x86_64-pc-linux-gnu".
Type "show configuration" for configuration details.
 
For information about how to find Technical Support, Product Updates,
User Forums, FAQs, tips and tricks, and other support information, please visit:
<[http://www.intel.com/software/products/support/>.](http://www.intel.com/software/products/support/%3E.)
For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from python...
(No debugging symbols found in python)
(gdb) r
Starting program: /home/gta/intel-xpu-backend-for-triton/.venv_debug/bin/python python/tutorials/02-fused-softmax.py
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
Registering SYCL extensions for gdb
[New Thread 0x7fff4afff6c0 (LWP 3361515)]
[New Thread 0x7fff4a7fe6c0 (LWP 3361516)]
[New Thread 0x7fff49ffd6c0 (LWP 3361517)]
[New Thread 0x7fff497fc6c0 (LWP 3361518)]
[New Thread 0x7fff48ffb6c0 (LWP 3361519)]
[New Thread 0x7fff487fa6c0 (LWP 3361520)]
[New Thread 0x7fff47ff96c0 (LWP 3361521)]
[New Thread 0x7fff477f86c0 (LWP 3361522)]
[New Thread 0x7fff46ff76c0 (LWP 3361523)]
[New Thread 0x7fff467f66c0 (LWP 3361524)]
[New Thread 0x7fff45ff56c0 (LWP 3361525)]
[New Thread 0x7fff457f46c0 (LWP 3361526)]
[New Thread 0x7fff44ff36c0 (LWP 3361527)]
[New Thread 0x7fff447f26c0 (LWP 3361528)]
[New Thread 0x7fff43ff16c0 (LWP 3361529)]
[New Thread 0x7fff437f06c0 (LWP 3361530)]
[New Thread 0x7fff42fef6c0 (LWP 3361531)]
[New Thread 0x7fff427ee6c0 (LWP 3361532)]
[New Thread 0x7fff41fed6c0 (LWP 3361533)]
[New Thread 0x7fff417ec6c0 (LWP 3361534)]
[New Thread 0x7fff40feb6c0 (LWP 3361535)]
[New Thread 0x7fff407ea6c0 (LWP 3361536)]
[New Thread 0x7fff3ffe96c0 (LWP 3361537)]
[New Thread 0x7fff3f7e86c0 (LWP 3361538)]
[New Thread 0x7fff3efe76c0 (LWP 3361539)]
[New Thread 0x7fff3e7e66c0 (LWP 3361540)]
[New Thread 0x7fff3dfe56c0 (LWP 3361541)]
[New Thread 0x7fff3d7e46c0 (LWP 3361542)]
[New Thread 0x7fff3cfe36c0 (LWP 3361543)]
[New Thread 0x7fff3c7e26c0 (LWP 3361544)]
[New Thread 0x7fff3bfe16c0 (LWP 3361545)]
[Detaching after vfork from child process 3361561]
[Detaching after vfork from child process 3361562]
[New Thread 0x7fff055ff6c0 (LWP 3361563)]
intelgt: gdbserver-ze started for process 3361512.
(I): Detected 3456 spills, recompiling the kernel using large GRF mode
(I): Kernel has now 3456 spills
(I): Detected 3456 spills for  "softmax_kernel"
(I): Detected 3584 spills, recompiling the kernel using large GRF mode
(I): Kernel has now 3584 spills
(I): Detected 3584 spills for  "softmax_kernel"
(I): Detected 3584 spills, recompiling the kernel using large GRF mode
(I): Kernel has now 3584 spills
(I): Detected 3584 spills for  "softmax_kernel"
(I): Detected 3584 spills, recompiling the kernel using large GRF mode
(I): Kernel has now 3584 spills
(I): Detected 3584 spills for  "softmax_kernel"
(I): Detected 3584 spills, recompiling the kernel using large GRF mode
(I): Kernel has now 3584 spills
(I): Detected 3584 spills for  "softmax_kernel"
(I): Detected 3584 spills, recompiling the kernel using large GRF mode
(I): Kernel has now 3584 spills
(I): Detected 3584 spills for  "softmax_kernel"
(I): Detected 3584 spills, recompiling the kernel using large GRF mode
(I): Kernel has now 3584 spills
(I): Detected 3584 spills for  "softmax_kernel"
(I): Detected 5248 spills, recompiling the kernel using large GRF mode
(I): Kernel has now 3584 spills
(I): Detected 3584 spills for  "softmax_kernel"
softmax-performance:
          N  Triton (GB/s)  Torch (GB/s)  Naive Softmax (GB/s)
0     256.0      41.560613    233.496393             76.548882
1     384.0      39.969066    245.776396            103.009327
2     512.0      54.667591    253.745812            112.416603
3     640.0      37.819742    266.993495             97.499607
4     768.0      45.884017    281.397159             91.849008
5     896.0      55.118259    290.245059             88.644924
6    1024.0      64.456885    302.517543             88.200057
7    1152.0      37.404522    311.750640             85.936574
8    1280.0      41.558044    318.230024             85.158925
9    1408.0      45.682625    320.482450             85.241556
10   1536.0      49.706329    324.828675             81.310449
11   1664.0      53.890349    332.072299             85.641297
12   1792.0      58.340976    336.067329             86.647032
13   1920.0      62.631919    341.662534             87.321490
14   2048.0      67.001476    345.082062             87.628713
15   2176.0      35.569025    345.608987             88.007241
16   2304.0      37.607323    349.654352             88.400723
17   2432.0      39.670834    351.293597             89.188120
18   2560.0      41.679006    355.519746             89.602249
19   2688.0      43.703226    355.957145             89.281288
20   2816.0      45.751070    359.503696             90.213516
21   2944.0      47.754271    357.513829             90.657330
22   3072.0      49.684612    361.783847             76.871297
23   3200.0      51.740522    362.742448             91.433636
24   3328.0      53.737165    365.955263             91.670347
25   3456.0      55.827960    364.594066             91.962116
26   3584.0      57.798444    367.439720             91.787605
27   3712.0      59.831003    366.284894             92.510780
28   3840.0      61.900569    368.908931             92.303287
29   3968.0      63.842833    369.491657             92.809512
30   4096.0      65.920369    371.210772             93.538858
31   4224.0      64.615394    371.597077             93.143391
32   4352.0      66.470632    372.171222             93.643212
33   4480.0      68.270454    375.512193             93.384257
34   4608.0      70.128890    372.992972             91.257677
35   4736.0      71.883612    374.944380             93.929297
36   4864.0      73.657499    372.919711             94.103815
37   4992.0      75.347490    375.673761             94.378235
38   5120.0      77.156592    375.819748             94.757582
39   5248.0      78.933433    377.363035             94.499869
40   5376.0      80.629531    377.554215             94.614386
41   5504.0      82.347196    379.027680             94.912858
42   5632.0      84.129637    376.768079             95.091163
43   5760.0      85.725204    381.044510             94.973432
44   5888.0      87.587638    378.403780             95.294603
45   6016.0      89.237804    382.239524             95.372103
46   6144.0      91.418280    381.134955             86.568757
47   6272.0      92.672830    383.507147             95.517714
48   6400.0      94.270983    381.202486             95.836623
49   6528.0      95.788293    383.659120             95.532698
50   6656.0      97.352123    380.757502             95.903645
51   6784.0      98.848256    383.460954             95.691277
52   6912.0     100.271373    380.973914             95.919724
53   7040.0     101.983076    383.571732             95.841833
54   7168.0     103.870238    382.097389             96.182970
55   7296.0     105.289545    384.154346             96.208387
56   7424.0     106.948771    384.126329             96.384357
57   7552.0     108.594232    386.001603             96.371059
58   7680.0     110.267731    385.008189             94.929429
59   7808.0     111.928985    385.382867             96.411827
60   7936.0     113.581510    383.825817             96.335330
61   8064.0     114.779515    384.235449             96.557125
62   8192.0     117.363996    383.428424             96.686003
63   8320.0     108.194729    383.196644             96.789227
64   8448.0     109.632799    386.735449             96.428940
65   8576.0     111.077989    384.272005             96.716902
66   8704.0     112.932520    387.286214             96.643077
67   8832.0     114.074819    383.782012             96.816366
68   8960.0     115.879025    386.899722             96.692077
69   9088.0     117.277490    384.678385             97.068728
70   9216.0     119.089880    387.023319             90.775628
71   9344.0     120.389434    384.884833             97.080329
72   9472.0     122.085222    387.100957             97.228345
73   9600.0     123.524893    385.045212             97.242239
74   9728.0     125.222870    386.458609             97.150968
75   9856.0     126.370193    384.722945             97.311058
76   9984.0     127.871567    387.248861             97.261455
77  10112.0     129.240383    385.184046             97.466458
78  10240.0     130.933312    387.396301             97.315664
79  10368.0     132.231707    384.851243             97.529641
80  10496.0     133.568533    387.974195             97.398791
81  10624.0     134.902294    385.506943             97.683565
82  10752.0     136.302926    387.708476             96.199803
83  10880.0     137.654206    386.464712             97.472085
84  11008.0     139.042496    388.216932             97.489052
85  11136.0     140.482487    387.096765             97.560059
86  11264.0     142.207975    388.899158             97.540280
87  11392.0     143.179004    387.167813             97.663622
88  11520.0     144.415406    388.359896             97.594853
89  11648.0     145.622684    387.196417             97.827307
90  11776.0     147.161616    388.358106             97.901706
91  11904.0     148.323524    389.288035             97.576072
92  12032.0     149.760912    388.112936             97.969084
93  12160.0     151.147185    389.397186             97.929250
94  12288.0     152.845809    389.011507             92.816492
95  12416.0     153.725660    389.133493             98.079793
96  12544.0     155.027532    388.828604             97.997668
97  12672.0     156.263004    389.363733             97.805690
[Thread 0x7fff3c7e26c0 (LWP 3361544) exited]
[Thread 0x7fff3cfe36c0 (LWP 3361543) exited]
[Thread 0x7fff3d7e46c0 (LWP 3361542) exited]
[Thread 0x7fff3efe76c0 (LWP 3361539) exited]
[Thread 0x7fff3ffe96c0 (LWP 3361537) exited]
[Thread 0x7fff417ec6c0 (LWP 3361534) exited]
[Thread 0x7fff3f7e86c0 (LWP 3361538) exited]
[Thread 0x7fff407ea6c0 (LWP 3361536) exited]
[Thread 0x7fff3dfe56c0 (LWP 3361541) exited]
[Thread 0x7fff3e7e66c0 (LWP 3361540) exited]
[Thread 0x7fff40feb6c0 (LWP 3361535) exited]
[Thread 0x7fff41fed6c0 (LWP 3361533) exited]
[Thread 0x7fff427ee6c0 (LWP 3361532) exited]
[Thread 0x7fff42fef6c0 (LWP 3361531) exited]
[Thread 0x7fff437f06c0 (LWP 3361530) exited]
[Thread 0x7fff43ff16c0 (LWP 3361529) exited]
[Thread 0x7fff447f26c0 (LWP 3361528) exited]
[Thread 0x7fff44ff36c0 (LWP 3361527) exited]
[Thread 0x7fff457f46c0 (LWP 3361526) exited]
[Thread 0x7fff45ff56c0 (LWP 3361525) exited]
[Thread 0x7fff467f66c0 (LWP 3361524) exited]
[Thread 0x7fff46ff76c0 (LWP 3361523) exited]
[Thread 0x7fff477f86c0 (LWP 3361522) exited]
[Thread 0x7fff47ff96c0 (LWP 3361521) exited]
[Thread 0x7fff487fa6c0 (LWP 3361520) exited]
[Thread 0x7fff48ffb6c0 (LWP 3361519) exited]
[Thread 0x7fff497fc6c0 (LWP 3361518) exited]
[Thread 0x7fff49ffd6c0 (LWP 3361517) exited]
[Thread 0x7fff4a7fe6c0 (LWP 3361516) exited]
[Thread 0x7fff4afff6c0 (LWP 3361515) exited]
[Thread 0x7fff3bfe16c0 (LWP 3361545) exited]
[Thread 0x7fff055ff6c0 (LWP 3361563) exited]
[Inferior 1 (process 3361512) exited normally]
Detached from device [0000:03:00.0]
[Inferior 2 (device [0000:03:00.0]) detached]
intelgt: inferior 2 (gdbserver-ze) has been removed.
(gdb) q

**Comment by radocaj-int** [comment]
even runcontrol with gdb works with 2026.0 👍

Thread 2.1121 hit Breakpoint 1, with SIMD lanes [0-31], softmax_kernel (output_ptr=0xffffffffff800800, input_ptr=0xffffffffff800000, input_row_stride=16, output_row_stride=16, n_rows=32, n_cols=16) at 02-fused-softmax.py:88
88          row_start = tl.program_id(0)
(gdb) l  
83
84      @triton.jit
85      def softmax_kernel(output_ptr, input_ptr, input_row_stride, output_row_stride, n_rows, n_cols,
86                         BLOCK_SIZE: tl.constexpr):
87          # starting row of the program
88          row_start = tl.program_id(0)
89          row_step = tl.num_programs(0)
90          for row_idx in tl.range(row_start, n_rows, row_step):
91              # The stride represents how much we need to increase the pointer to advance 1 row
92              row_start_ptr = input_ptr + row_idx * input_row_stride
(gdb) set scheduler-locking on
(gdb) n
89          row_step = tl.num_programs(0)
(gdb) n
92              row_start_ptr = input_ptr + row_idx * input_row_stride
(gdb) s
95              col_offsets = tl.arange(0, BLOCK_SIZE)
(gdb) s
98              mask = col_offsets < n_cols
(gdb) s
107             output_row_start_ptr = output_ptr + row_idx * output_row_stride
(gdb) p mask
$1 = <optimized out>
(gdb) bt
#0  softmax_kernel (output_ptr=0xffffffffff800800, input_ptr=0xffffffffff800000, input_row_stride=16, output_row_stride=16, n_rows=32, n_cols=16) at 02-fused-softmax.py:107
(gdb) p n_cols                
$2 = 16

-> we can close this ticket.

**Comment by anmyachev** [comment]
@radocaj-int thanks a lot!

With 2026.0 it should work at least starting from Triton 3.7.2.

Closing it.
